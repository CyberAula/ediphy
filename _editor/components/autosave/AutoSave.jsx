import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Ediphy from '../../../core/editor/main';
import { CHANGE_DISPLAY_MODE, EXPAND_NAV_ITEM, IMPORT_STATE, INCREASE_LEVEL, INDEX_SELECT, SELECT_BOX, SELECT_NAV_ITEM, SET_BUSY, TOGGLE_TEXT_EDITOR, TOGGLE_TITLE_MODE, UPDATE_NAV_ITEM_EXTRA_FILES, UPDATE_BOX } from './../../../common/actions';
const lang = ['en', 'es'];
/** *
 * Component for auto-saving the state of the application periodically and avoid losing changes
 */
export default class AutoSave extends Component {
    /**
     * Constructor
     * @param props React component props
     */
    constructor(props) {
        super(props);
        /** *
         * Component's initial state
         * @type {{displaySave: boolean Info displayed to user}}
         */
        this.state = {
            displaySave: false,
            modifiedState: false,
        };
    }

    /**
     * After component mounts
     * Sets up timer for autosaving
     */
    componentDidMount() {
        /** *
         * Timer function set up
         * @type {js timer}
         */
        this.intervalId = setInterval(this.timer.bind(this), Ediphy.Config.autosave_time);
    }

    /**
     * Before component unmounts
     * Clears timer
     */
    componentWillUnmount() {
        // use intervalId from the state to clear the interval
        clearInterval(this.intervalId);
    }

    /**
     * Before component receives props
     * Displays message to user for 2 seconds
     * @param nextProps
     */
    componentWillReceiveProps(nextProps) {
        if (nextProps.isBusy.value) {
            this.setState({ displaySave: true });
            setTimeout(() => {
                this.setState({ displaySave: false });
            }, 2000);
        }

        if (this.state.modifiedState === false) {
            switch(nextProps.lastAction) {
            case CHANGE_DISPLAY_MODE:
            case EXPAND_NAV_ITEM:
            case IMPORT_STATE:
            case INCREASE_LEVEL:
            case INDEX_SELECT:
            case SELECT_BOX:
            case SELECT_NAV_ITEM:
            case SET_BUSY:
            case TOGGLE_TEXT_EDITOR:
            case TOGGLE_TITLE_MODE:
            case UPDATE_NAV_ITEM_EXTRA_FILES:
            case UPDATE_BOX:
                return;
            default:
                this.setState({ modifiedState: true });
            }
        }
    }

    /**
     * Function run when time is up. It saves changes.
     */
    timer() {
        if(!this.props.visorVisible && this.state.modifiedState) {
            this.props.save();
        }
    }

    /**
     * Renders React Component
     * @returns {code} React rendered component
     */
    render() {
        return(
            <div className="savingLabel"
                style={{ display: this.state.displaySave ? 'block' : 'none' }}>{'Guardando...'}
            </div>
        );
    }
}

AutoSave.propTypes = {
    /**
   * Saves the state
   */
    save: PropTypes.func.isRequired,
    /**
   * Last Redux action dispatched
   */
    lastAction: PropTypes.any,
    /**
   * Indicates if there is a current server operation
   */
    isBusy: PropTypes.any,
    /**
   * Indicates if the preview mode is on or not
   */
    visorVisible: PropTypes.bool,

};
