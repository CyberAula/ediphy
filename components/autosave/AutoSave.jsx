import React, {Component} from 'react';
import Dali from './../../core/main';

export default class AutoSave extends Component {

    constructor(props){
        super(props);
        this.state = {displaySave: false};
    }

    componentDidMount() {
        this.intervalId = setInterval(this.timer.bind(this), Dali.Config.autosave_time);
    }

    componentWillUnmount() {
        // use intervalId from the state to clear the interval
        clearInterval(this.intervalId);
    }

    componentWillReceiveProps(nextProps){

        if(nextProps.isBusy.value){
            this.setState({displaySave: true});
            setTimeout(() => {
                this.setState({displaySave: false});
            }, 2000);
        }
    }

    timer() {
        this.props.save();
    }

    render(){
        return(
            /* jshint ignore:start */
            <div className="savingLabel"
                style={{display: this.state.displaySave ? 'block' : 'none'}}>{'Guardando...'}</div>
            /* jshint ignore:end */
        );
    }
}