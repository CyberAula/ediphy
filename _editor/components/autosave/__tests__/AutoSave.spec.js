import React from 'react';
import expect from 'expect';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import AutoSave from "../AutoSave.jsx";

describe('Autosave', () => {
    it('should render with isRequired props', () => {
        const saving = shallow(
            <AutoSave save={() => { console.log("Saving..."); }}/>);
        expect(shallowToJson(saving)).toMatchSnapshot();
    });
});
