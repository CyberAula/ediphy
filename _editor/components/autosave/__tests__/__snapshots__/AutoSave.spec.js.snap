// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autosave should render with isRequired props 1`] = `
<div
  className="savingLabel"
  style={
    Object {
      "display": "none",
    }
  }
>
  Guardando...
</div>
`;
